Contributing with examples

pjpmarques opened the pull request:
Happy to contribute with a few jupyter notebooks on how to use the package.
You can check the examples/result here:
* [Simple differential equation](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Simple_Differential_Equation.ipynb)
* [Lorenz Attractor](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Lorenz_Attractor.ipynb)
* [Terminal Velocity](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Terminal_Velocity.ipynb)

mauro3 commented on README.md:
```diff
@@ -42,6 +42,13 @@ Additionally, `ode23s` solver supports
+* [Simple differential equation](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Simple_Differential_Equation.ipynb)
+* [Lorenz Attractor](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Lorenz_Attractor.ipynb)
+* [Terminal Velocity](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Terminal_Velocity.ipynb)
+
```
These links would need to point to this repo and not your fork.

tlycken replied:
It might be (but isn't necessarily) nice to first merge this, though, and then when the notebooks exist in ODE.jl, refer to them with a specific hash rather than `master`. This would ensure that links aren't broken by mistake if/when the notebooks are (re)moved, but on the other hand it would require manual update of all links when the notebooks are changed, even if the location did not change; and furthermore, that manual change would not be possible to make before the changes to the notebooks have been committed.

mauro3 replied:
Ideally the notebooks would be checked with each commit. Is this possible?

tlycken replied:
"Checked" as in "run-through with no errors"?

No idea - I doubt that it's easy. It would be a nice feature of IJulia, though, so if it doesn't exist it's worth a feature request there. I imagine something like this:

```
using IJulia
run_all_cells("path/to/notebook.ipynb")
```

mauro3 commented on examples/README.md:
```diff
@@ -0,0 +1,10 @@
+You can also directly see the result without installing anything. Just follow the links:
+* [Simple differential equation](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Simple_Differential_Equation.ipynb)
+* [Lorenz Attractor](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Lorenz_Attractor.ipynb)
+* [Terminal Velocity](http://nbviewer.jupyter.org/github/pjpmarques/ODE.jl/blob/master/examples/Terminal_Velocity.ipynb)
```
Again the links.

mauro3 commented on examples/Simple_Differential_Equation.ipynb:
```diff
@@ -0,0 +1,194 @@
+   },
+   "outputs": [],
+   "source": [
+    "using PyPlot\n",
```
Why not do the `using PyPlot` later, just before the plotting starts?

mauro3 commented on examples/Simple_Differential_Equation.ipynb:
```diff
@@ -0,0 +1,194 @@
+    "\n",
+    "This notebook illustrates how to solve a simple differential equation using [Julia's ODE library](https://github.com/JuliaLang/ODE.jl).\n",
+    "\n",
+    "The equation $ x''(t) + x(t) == 0 $ has as possible solution $ x(t) = sin(t) $\n",
```
I would not use the `==`, that is not maths notation.

mauro3 commented on examples/Simple_Differential_Equation.ipynb:
```diff
@@ -0,0 +1,194 @@
+   },
+   "outputs": [],
+   "source": [
+    "t, y = ode23(f, start, time);"
```
I'd use `ode45`, that is probably the standard non-stiff solver.

mauro3 commented on examples/Simple_Differential_Equation.ipynb:
```diff
@@ -0,0 +1,194 @@
+    "start = [0.0; 0.1]\n",
+    "\n",
+    "# Time vector going from 0 to 2*PI in 0.01 increments\n",
+    "time = collect(0:0.1:4*pi);"
```
Do not use `collect` here, it is not necessary.

mauro3 commented on examples/Terminal_Velocity.ipynb:
```diff
@@ -0,0 +1,295 @@
+    "This notebook discusses the <a href=\"http://en.wikipedia.org/wiki/Drag_(physics)\">drag forces</a> exerted on a body when traveling through air. It's also done in [Julia](julialang.org) so that I can better understand the potencial of the language.\n",
+    "\n",
+    "![](http://upload.wikimedia.org/wikipedia/en/8/89/Parachuters_in_hybrid_formation.jpg)  \n",
+    "\n",
```
Give attribution

mauro3 commented on examples/Terminal_Velocity.ipynb:
```diff
@@ -0,0 +1,295 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A falling body is subject to two forces: a downward force, $\\vec{F_g}$, due to gravity, and an upward force $\\vec{F_a}$, due to air resistance. Let's consider a body with mass $m$. Assuming an upward oriented YY axis, the following applies:\n",
```
`YY` -> `Y`

mauro3 commented on examples/Terminal_Velocity.ipynb:
```diff
@@ -0,0 +1,295 @@
+    "\n",
+    "---\n",
+    "\n",
+    "So, let's simulate this with Julia.\n",
```
No need for this sentence.

mauro3 commented on examples/Terminal_Velocity.ipynb:
```diff
@@ -0,0 +1,295 @@
+   "outputs": [],
+   "source": [
+    "# Constants \n",
+    "g  = 9.8                # Accelaration of gravity\n",
```
It is bad style and bad for performance to use non-const globals.  Instead use `const g =9.8` etc.

mauro3 commented on examples/Terminal_Velocity.ipynb:
```diff
@@ -0,0 +1,295 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, that's what I call pretty neat!"
```
Maybe a bit too colloquial for docs?

mauro3 commented:
Thanks for these examples!  I think the would be good to have in ODE.jl.

At the moment there are no docs for ODE.jl.  But one thing to ponder is how we'd like to have examples.  Are notebooks a suitable format?  However, if the answer is "no", then these could be ported over.

A few comments for the Lorenz notebook (for some reason, I cannot do them inline):

- put a link to wikipedia in the intro?
- remove "Let's implement it in Julia."
- make your global `sigma`, `rho`, `bet` all `const`
- any reason to not use `ode45`?

Also, squash the commits into one.

pjpmarques commented:
Hi. Glad to correct the code and follow the suggestions. I've:
- Solved the links problem pointing to outside repository. 
- Marked constants with "const" 
- Gave credit to Wikimedia and include references to wikipedia 
- Changed ode23() to ode45() 
- Removed personal credit 
- Corrected some incorrect notation (use of == instead of =; use of YY instead of Y) 
- Use of less colloquial style
- Squashed all commits into one

pjpmarques commented:
Hi. Any news on this and/or do you want me to correct something? Thanks.

mauro3 commented:
LGTM
